README.md
OpenBCI Hub Application
==============

A minimalistic yet comprehensive boilerplate application for [Electron runtime](http://electron.atom.io). Tested on OSX, Windows and Linux.  

This project does not impose on you any framework (like Angular or React). Instead, it tries to give you only the 'electron' part of technology stack so you can pick your favorite tools for the rest.

# Quick start
The only development dependency of this project is [Node.js](https://nodejs.org). So just make sure you have it installed.
Then type few commands known to every Node developer...
```
git clone https://github.com/OpenBCI/OpenBCI_Hub.git
cd OpenBCI_Hub
npm install
npm install electron-prebuilt -g
npm start
```
... and boom! You have a running desktop application on your screen.

**On Windows:** open a PowerShell and run:
```
npm install --global --production windows-build-tools
```
If you have any issues, try:
```
npm install --build-from-source
```
# Structure of the project

## Declaring dependencies

There are **two** `package.json` files:

#### 1. `package.json` for development
Sits on path: `electron-boilerplate/package.json`. This is where you should declare dependencies for your development environment and build scripts. **This file is not distributed with real application!**

It's also the place to specify the Electron runtime version you want to use:
```json
"devDependencies": {
  "electron": "1.8.4"
}
```
Note: [Electron authors advise](http://electron.atom.io/docs/tutorial/electron-versioning/) to use fixed version here.

#### 2. `package.json` for your application
Sits on path: `electron-boilerplate/app/package.json`. This is **real** manifest of your application. Declare your app dependencies here.

#### OMG, but seriously why there are two `package.json`?
1. Native npm modules (those written in C, not JavaScript) need to be compiled, and here we have two different compilation targets for them. Those used in application need to be compiled against electron runtime, and all `devDependencies` need to be compiled against your locally installed node.js. Thanks to having two files this is trivial.
2. When you package the app for distribution there is no need to add up to size of the app with your `devDependencies`. Here those are always not included (reside outside the `app` directory).

## Folders for application code

The application is split between two main folders...

`src` - this folder is intended for files which need to be transpiled or compiled (files which can't be used directly by electron).

`app` - contains all static assets (put here images, css, html etc.) which don't need any pre-processing.

The build process compiles all stuff from the `src` folder and puts it into the `app` folder, so after the build has finished, your `app` folder contains the full, runnable application.

Treat `src` and `app` folders like two halves of one bigger thing.

The drawback of this design is that `app` folder contains some files which should be git-ignored and some which shouldn't (see `.gitignore` file). But thanks to this two-folders split development builds are much (much!) faster.

# Development

### Installation

```
npm install
```
It will also download Electron runtime and install dependencies for the second `package.json` file inside the `app` folder.

### Starting the app

```
npm start
```

### Adding npm modules to your app

Remember to add your dependencies to `app/package.json` file:
```
cd app
npm install name_of_npm_module --save
```

### Working with modules

Thanks to [rollup](https://github.com/rollup/rollup) you can (and should) use ES6 modules for all code in `src` folder. But because ES6 modules still aren't natively supported you can't use them in the `app` folder.

Use ES6 syntax in the `src` folder like this:
```js
import myStuff from './my_lib/my_stuff';
```

But use CommonJS syntax in `app` folder. So the code from above should look as follows:
```js
var myStuff = require('./my_lib/my_stuff');
```

# Testing

### Unit tests

Using [electron-mocha](https://github.com/jprichardson/electron-mocha) test runner with the [chai](http://chaijs.com/api/assert/) assertion library. To run the tests go with standard and use the npm test script:
```
npm test
```
This task searches for all files in `src` directory which respect pattern `*.spec.js`.

### End to end tests

Using [mocha](https://mochajs.org/) test runner and [spectron](http://electron.atom.io/spectron/). Run with command:
```
npm run e2e
```
This task searches for all files in `e2e` directory which respect pattern `*.e2e.js`.

### Code coverage

Using [istanbul](http://gotwarlost.github.io/istanbul/) code coverage tool. Run with command:
```
npm run coverage
```
You can set the reporter(s) by setting `ISTANBUL_REPORTERS` environment variable (defaults to `text-summary` and `html`). The report directory can be set with `ISTANBUL_REPORT_DIR` (defaults to `coverage`).

### Continuous integration

Electron [can be plugged](https://github.com/atom/electron/blob/master/docs/tutorial/testing-on-headless-ci.md) into CI systems. Here two CIs are preconfigured for you. [Travis CI](https://travis-ci.org/) covers testing on OSX and Linux and [App Veyor](https://www.appveyor.com) on Windows.

# Making a release

To package your app into an installer use command:
```
npm run release
```
It will start the packaging process for operating system you are running this command on. Ready for distribution file will be outputted to `dist` directory.

You can create Windows installer only when running on Windows, the same is true for Linux and OSX. So to generate all three installers you need all three operating systems.

All packaging actions are handled by [electron-builder](https://github.com/electron-userland/electron-builder). See docs of this tool if you want to customize something.

**Note:** There are various icons and bitmap files in `resources` directory. Those are used in installers and intended to be replaced by your own graphics.

**Note:** After OpenBCI NPM package update (ex. [@openbci/cyton](https://www.npmjs.com/package/@openbci/cyton)), use `npm run clean-install` and then `npm run release`.

# License

Released under the MIT license.


CHANGELOG.md
# v2.1.0

### Bug Fixes

- Return cyton firmware version in wifi connect response

# v2.0.9

### Bug Fixes

- Fix BLED112 Impedance check by allowing 2 seconds for command/success OpenBCI/OpenBCI_GUI#421

# v2.0.8

### Bug Fixes

- Get host IP address on Windows machines using valid network interface

# v2.0.7

### Bug Fixes

- Display Cyton firmware in GUI OpenBCI/OpenBCI_GUI#449
- Cyton serial SD card mode causes GUI init to fail OpenBCI/OpenBCI_GUI#386

# v2.0.6

### New Features

- Add "clean-install" script to package.json

### Bug Fixes

- Update OpenBCI npm module names
- Fix Mac built-in BLE functionality

# v2.0.5

### Bug Fixes

- Cyton impedance check functionality

# v2.0.4

### New Features

- Added version number to the HUB's display

# v2.0.3

### Bug Fixes

- Ganglion was not working on Mojave

# v2.0.2

### Bug Fixes

- Clean up logs

# v2.0.1

### Bug Fixes

- Fix bugs with accelerometer, BLED112, and impedance

# v2.0.0

### Breaking Changes

- Now using JSON to send and recieve data from the HUB!

### New Features

- Started sending time stamps with each data sample

# v1.4.6

### New Features

- Bumped openbci-ganglion to 1.1.12 for BLED112 support for linux

# v1.4.5

### New Features

- Bumped openbci-ganglion to 1.1.11 for BLED112 support for windows

# v1.4.4

### Bug Fixes

- Bumped openbci-ganglion to 1.1.9 for accel patches

# v1.4.3

### New Features

- Bumped Electron to 2.0.2

# v1.4.2

### New Features

- Working with static ip

# v1.4.1

### Bug Fixes

- Ganglion with BLED112 was not working to discover devices, bumped to v1.1.7

# v1.4.0

### New Features

- Add BLED112 support via OpenBCI Ganglion 1.1.5

### Chores

- Bumped cyton to 1.1.1 for new serialport
- Bumped electron to 1.8.2

# v1.3.9

### Bug Fixes

- UDP Burst did not work because was sending to `/udpBurst` instead of `/udp`
- App would not close on uncaught exceptions, now, an error box is shown with the error, then when the user hit's ok, the whole app quits.

# v1.3.8

### Bug Fixes

- Found out that `openbci-ganglion` had it's own version of noble it was using that was not the macos high seirra updated one, so i copy and pasted the correct build into both app/node_moduloes/noble AND app/node_modules/openbci-ganglion/node_modules/noble. More on this issue can be found at [openbci/openbci_gui/issue/270](https://github.com/OpenBCI/OpenBCI_GUI/issues/270)

# v1.3.7

### Bug Fixes

- WiFi would send success message on 404 errors. Now sends error code 435, or, update your wifi shield firmware to support this feature.

# v1.3.6

### Bug Fixes

- Application in production was not finding custom OpenBCI logo.
- Process command error would send message type error to GUI for all boards. Changes to send command type with error code.

# v1.3.5

### Bug Fixes

- Daisy data did not send aux values

# v1.3.3/4

### Bug Fixes

- Update ganglion node driver to 1.0.0
- Stopped wifi scan in wifi cleanup
- Cleaned up event listeners for cyton/ganglion/wifi disconnect
- Fixed bug with daisy not getting accel data or stop byte by bumping wifi version to 0.3.0
- Daisy with cyton now get's stop bytes with bump to 1.0.6

# v1.3.2

### Bug Fixes

- SD card did not work for wifi on cyton

# v1.3.1

### Bug Fixes

- Removed annoying pop-up on windows

# v1.3.0

### Bug Fixes

- Issue with ganglion channel data not sent
- Issue where cyton aux data not sent

### Breaking Changes

- Ganglion data over wifi has only 4 channels (as it's supposed to)
- Ganglion accel data over wifi sent with packet instead of in separate packet to prevent misalignment.

# v1.2.0

Fixing bugs with AppVeyor build service.

# v1.1.3

### Bug Fixes

- Fixed bugs with process protocol and several others.

# v1.1.2

Add a lot more fixes.

# v1.1.0

Add a lot more fixes.

# v1.0.2

Fix many issues with cyton and ganglion and wifi.

# v1.0.1

Add channel setting commands.

# v1.0.0

Add cyton and wifi support

# v0.4.1

### New Features

- BLE error on start up now sends error

### Bug Fixes

- Fixes #12 - Absorb 'no valid USB' found and send log

# v0.4.0

### Breaking Changes

- Changed name of built app from `Ganglion Hub` to `GanglionHub`.

### Bug Fixes

- On client leave if ganglion is connected, the connection will close.

# v0.3.1

### Enhancements

- Building the proper builds by tweaking appveryor.

# v0.3.0

### Enhancements

- Standardization of Specification.

### Breaking changes

- Accelerometer, Impedance, and Sample data all have specific success error codes: 202, 203, and 204 respectively. Prior to this version all were using the same 200 code.

# v0.2.3

### Enhancements

- Calling connect with device name now performs a scan to ensure that device is really still available to connect to.

### Bug Fixes

- Calling connect with timeout caused another bug, for timeout.

# v0.2.2

### Enhancements

- Bump `openbci-ganglion` to `0.4.1`.
- Calling connect now has a timeout!

### Bug Fixes

- Dropped connections now eject a message out to connected client.

# v0.2.1

### Bug Fixes

- `ganglionFound` event emitter was not removed on start of new scan.

### Enhancements

- Disabled verbose print out for production build.
- Bump `openbci-ganglion` to `0.3.8`.

# v0.2.0

### Bug Fixes

- Disconnect did not clean up event emitters added in connect.

### Enhancements

- Bump `openbci-ganglion` to `0.3.7`

# v0.1.6

### Enhancements

- Bump `openbci-ganglion` to `0.3.6`

### Bug Fixes

- Ganglion would not disconnect.
- Change Appveyor to Node 6

# v0.1.5

### Enhancements

- Bump `openbci-ganglion` to `0.3.3`

### Bug Fixes

- Ganglion could not stop seaching.

# v0.1.4

### New Features

- Add Accel

### Bug Fixes

- Ganglion could not connect twice.

# v0.1.3

### Bug Fixes

- Add accelerometer data flow
- Bump ganglion node to `0.3.0`

# v0.1.2

### Bug Fixes

- Fix bug with undefined impedance

# v0.1.1

### Enhancements

- Update to use 18 bit compression.
- Update to v0.2.0 of `openbci-ganglion`.
- Fix bug in impedance sending.


tests/status.py
#!/usr/bin/env python

import socket


TCP_IP = '127.0.0.1'
TCP_PORT = 10996
BUFFER_SIZE = 1024
MSGSREPLIES = [
        ("", "{"code":200,"type":"status"}\r\n"),
        ("q,sup,;\n", "q,200,;\n")
]

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect((TCP_IP, TCP_PORT))
for msgreply in MSGSREPLIES:
    s.send(msgreply[0])
    print "sent data:", repr(msgreply[0])
    data = s.recv(BUFFER_SIZE)
    print "received data:", repr(data)
    if data != msgreply[1]:
     raise AssertionError("Expected {} but got {}".format(repr(msgreply[1]), repr(data)))
s.close()



tasks/build_app.js
'use strict';

var gulp = require('gulp');
var less = require('gulp-less');
var watch = require('gulp-watch');
var batch = require('gulp-batch');
var plumber = require('gulp-plumber');
var jetpack = require('fs-jetpack');
var bundle = require('./bundle');
var utils = require('./utils');

var projectDir = jetpack;
var srcDir = jetpack.cwd('./src');
var destDir = jetpack.cwd('./app');

gulp.task('bundle', function () {
    return Promise.all([
        bundle(srcDir.path('background.js'), destDir.path('background.js')),
        bundle(srcDir.path('index.js'), destDir.path('index.js'))
        // bundle(srcDir.path('app.js'), destDir.path('app.js')),
    ]);
});

gulp.task('less', function () {
    return gulp.src(srcDir.path('stylesheets/main.less'))
        .pipe(plumber())
        .pipe(less())
        .pipe(gulp.dest(destDir.path('stylesheets')));
});

gulp.task('environment', function () {
    var configFile = 'config/env_' + utils.getEnvName() + '.json';
    projectDir.copy(configFile, destDir.path('env.json'), { overwrite: true });
    projectDir.copy('resources/icons/icon.png', destDir.path('resources/icons/icon.png'), { overwrite: true });
    // projectDir.copy('resources/icons/icon@2x.png', destDir.path('resources/icons/icon@2x.png'), { overwrite: true });
});

gulp.task('watch', function () {
    var beepOnError = function (done) {
        return function (err) {
            if (err) {
                utils.beepSound();
            }
            done(err);
        };
    };

    watch('src/**/*.js', batch(function (events, done) {
        gulp.start('bundle', beepOnError(done));
    }));
    watch('src/**/*.less', batch(function (events, done) {
        gulp.start('less', beepOnError(done));
    }));
});

gulp.task('build', ['bundle', 'less', 'environment']);


src/app.js
// Here is the starting point for your application code.
// All stuff below is just to show you how it works. You can delete all of it.

// Use new ES6 modules syntax for everything.
import os from 'os'; // native node.js module
import net from 'net';
import { remote } from 'electron'; // native electron module
// import jetpack from 'fs-jetpack'; // module loaded from npm
import { Ganglion, k } from 'openbci'; // native npm module
// import { greet } from './hello_world/hello_world'; // code authored by you in this project
import env from './env';
import * as _ from 'lodash';

console.log('Loaded environment variables:', env);

// const app = remote.app;
// const appDir = jetpack.cwd(app.getAppPath());

// Holy crap! This is browser window with HTML and stuff, but I can read
// here files like it is node.js! Welcome to Electron world :)
// console.log('The author of this app is:', appDir.read('package.json', 'json').author);
//
// document.addEventListener('DOMContentLoaded', function () {
//   document.getElementById('greet').innerHTML = greet();
//   document.getElementById('platform-info').innerHTML = os.platform();
//   document.getElementById('env-name').innerHTML = env.name;
// });

/** TCP */
const kTcpActionStart = 'start';
const kTcpActionStatus = 'status';
const kTcpActionStop = 'stop';
const kTcpCmdConnect = 'c';
const kTcpCmdCommand = 'k';
const kTcpCmdData = 't';
const kTcpCmdDisconnect = 'd';
const kTcpCmdError = 'e';
const kTcpCmdImpedance = 'z';
const kTcpCmdLog = 'l';
const kTcpCmdScan = 's';
const kTcpCmdStatus = 'q';
const kTcpCodeBadPacketData = 500;
const kTcpCodeSuccess = 200;
const kTcpCodeGanglionFound = 201;
const kTcpCodeStatusConnected = 300;
const kTcpCodeStatusDisconnected = 301;
const kTcpCodeStatusScanning = 302;
const kTcpCodeStatusNotScanning = 303;
const kTcpCodeErrorUnknown = 499;
const kTcpCodeErrorAlreadyConnected = 408;
const kTcpCodeErrorCommandNotRecognized = 406;
const kTcpCodeErrorDeviceNotFound = 405;
const kTcpCodeErrorNoOpenBleDevice = 400;
const kTcpCodeErrorUnableToConnect = 402;
const kTcpCodeErrorUnableToDisconnect = 401;
const kTcpCodeErrorScanAlreadyScanning = 409;
const kTcpCodeErrorScanNoneFound = 407;
const kTcpCodeErrorScanNoScanToStop = 410;
const kTcpCodeErrorScanCouldNotStart = 412;
const kTcpCodeErrorScanCouldNotStop = 411;
const kTcpHost = '127.0.0.1';
const kTcpPort = 10996;
const kTcpStop = ',;\n';

let verbose = true;
let ganglion = new Ganglion({
  nobleScanOnPowerOn: false,
  verbose: verbose
});

let _client = null;

// Start a TCP Server
net.createServer((client) => {
  // Identify this client
  client.name = `${client.remoteAddress}:${client.remotePort}`;

  if (verbose) {
    console.log(`Welcome ${client.name}`);
  }

  // Put this new client in the list
  if (_.isNull(_client)) {
    _client = client;
    console.log('client set');

    // Handle incoming messages from clients.
    client.on('data', (data) => {
      if (verbose) {
        console.log(`server got: ${data} from ${client.name}`);
      }
      parseMessage(data, client);
    });

    // Remove the client from the list when it leaves
    client.on('end', () => {
      _client = null;
      if (verbose) {
        console.log(`${client.name} has left`);
      }
      client.removeAllListeners('data');
      client.removeAllListeners('end');
      client.removeAllListeners('error');
    });

    client.on('error', (err) => {
      if (verbose) {
        console.log(`Error from ${client.name}: ${err}`);
      }
    });
  }
}).listen({
  port: kTcpPort,
  host: kTcpHost
});

if (verbose) {
  console.log(`server listening on port ${kTcpHost}:${kTcpPort}`);
}

var writeOutToConnectedClient = (message) => {
  if (_client) {
    _client.write(message);
  } else {
    throw Error('no connected client');
  }
};

/**
 * Called when a new sample is emitted.
 * @param sample {Object}
 *  A sample object.
 */
var sampleFunction = (sample) => {
  let packet = '';
  packet = `${kTcpCmdData},${kTcpCodeSuccess},`;
  packet += sample.sampleNumber;
  for (var j = 0; j < sample.channelData.length; j++) {
    packet += ',';
    packet += sample.channelData[j];
  }
  packet += `${kTcpStop}`;
  writeOutToConnectedClient(packet);
};

var parseMessage = (msg, client) => {
  let msgElements = msg.toString().split(',');
  // var char = String.fromCharCode(msg[0])
  // console.log('msgElements[0]',msgElements[0],char)
  switch (msgElements[0]) {
    case kTcpCmdConnect:
      if (ganglion.isConnected()) {
        if (verbose) console.log('already connected');
        client.write(`${kTcpCmdConnect},${kTcpCodeErrorAlreadyConnected}${kTcpStop}`);
      } else {
        if (verbose) console.log(`attempting to connect to ${msgElements[1]}`);
        ganglion.on(k.OBCIEmitterReady, () => {
          client.write(`${kTcpCmdConnect},${kTcpCodeSuccess}${kTcpStop}`);
          ganglion.on('sample', sampleFunction);
        });
        ganglion.connect(msgElements[1]) // Port name is a serial port name, see `.listPorts()`
          .catch((err) => {
            client.write(`${kTcpCmdConnect},${kTcpCodeErrorUnableToConnect},${err}${kTcpStop}`);
          });
      }
      break;
    case kTcpCmdCommand:
      if (ganglion.isConnected()) {
        ganglion.write(msgElements[1])
          .then(() => {
            if (verbose) console.log(`sent ${msgElements[1]} to Ganglion`);
            client.write(`${kTcpCmdCommand},${kTcpCodeSuccess}${kTcpStop}`);
          })
          .catch((err) => {
            if (verbose) console.log('sendCharacteristic not set');
            client.write(`${kTcpCmdError},${err}${kTcpStop}`);
          });
      } else {
        client.write(`${kTcpCmdCommand},${kTcpCodeErrorNoOpenBleDevice}${kTcpStop}`);
      }
      break;
    case kTcpCmdDisconnect:
      ganglion.manualDisconnect = true;
      ganglion.disconnect()
        .then(() => {
          client.write(`${kTcpCmdDisconnect},${kTcpCodeSuccess}${kTcpStop}`);
        })
        .catch((err) => {
          client.write(`${kTcpCmdDisconnect},${kTcpCodeErrorUnableToDisconnect},${err}${kTcpStop}`);
        });
      break;
    case kTcpCmdScan:
      processScan(msg, client);
      break;
    case kTcpCmdStatus:
      // A simple loop back
      client.write(`${kTcpCmdStatus},${kTcpCodeSuccess}${kTcpStop}`);
      break;
    case kTcpCmdError:
    default:
      client.write(`${kTcpCmdError},${kTcpCodeBadPacketData},Error: command not recognized${kTcpStop}`);
      break;
  }
};

const processScan = (msg, client) => {
  const ganglionFound = (peripheral) => {
    const localName = peripheral.advertisement.localName;
    client.write(`${kTcpCmdScan},${kTcpCodeGanglionFound},${localName}${kTcpStop}`);
  };
  let msgElements = msg.toString().split(',');
  const action = msgElements[1];
  switch (action) {
    case kTcpActionStart:
      if (ganglion.isSearching()) {
        client.write(`${kTcpCmdScan},${kTcpCodeErrorScanAlreadyScanning}${kTcpStop}`);
      } else {
        ganglion.on(k.OBCIEmitterGanglionFound, ganglionFound);
        ganglion.searchStart()
          .then(() => {
            client.write(`${kTcpCmdScan},${kTcpCodeSuccess},${kTcpActionStart}${kTcpStop}`);
          })
          .catch((err) => {
            ganglion.removeListener(k.OBCIEmitterGanglionFound, ganglionFound);
            client.write(`${kTcpCmdScan},${kTcpCodeErrorScanCouldNotStart},${err}${kTcpStop}`);
          });
      }
      break;
    case kTcpActionStatus:
      if (ganglion.isSearching()) {
        client.write(`${kTcpCmdScan},${kTcpCodeStatusScanning}${kTcpStop}`);
      } else {
        client.write(`${kTcpCmdScan},${kTcpCodeStatusNotScanning}${kTcpStop}`);
      }
      break;
    case kTcpActionStop:
      if (ganglion.isSearching()) {
        ganglion.searchStop()
          .then(() => {
            ganglion.removeListener(k.OBCIEmitterGanglionFound, ganglionFound);
            client.write(`${kTcpCmdScan},${kTcpCodeSuccess},${kTcpActionStop}${kTcpStop}`);
          })
          .catch((err) => {
            client.write(`${kTcpCmdScan},${kTcpCodeErrorScanCouldNotStop},${err}${kTcpStop}`);
          });
      } else {
        client.write(`${kTcpCmdScan},${kTcpCodeErrorScanNoScanToStop},${kTcpStop}`);
      }
      break;
  }
};

function exitHandler (options, err) {
  if (options.cleanup) {
    if (verbose) console.log('clean');
    // console.log(connectedPeripheral)
    ganglion.manualDisconnect = true;
    ganglion.disconnect();
  }
  if (err) console.log(err.stack);
  if (options.exit) {
    if (verbose) console.log('exit');
    if (ganglion.isConnected()) {
      ganglion.disconnect()
        .then(() => {
          process.exit();
        })
        .catch((err) => {
          if (verbose) console.log(err);
          process.exit();
        });
    }
  }
}

// do something when app is closing
process.on('exit', exitHandler.bind(null, {
  cleanup: true
}));

// catches ctrl+c event
process.on('SIGINT', exitHandler.bind(null, {
  exit: true
}));

// catches uncaught exceptions
process.on('uncaughtException', exitHandler.bind(null, {
  exit: true
}));
